xmodem3.4-unix/xmodem.1
.TH XMODEM LOCAL "October 4, 1987"
.UC 4.2
.SH NAME
xmodem \- Christensen protocol file transfer utility
.SH SYNOPSIS
.B xmodem
[\fBst|sb|rt|rb\fR][\fBymkcdlx\fR]
[file...]
.br
.SH DESCRIPTION
The
.I xmodem
program implements the Christensen (XMODEM) file transfer
protocol for moving files between 4.2/4.3BSD Unix systems and microcomputers.
The XMODEM/CRC protocol, the MODEM7 batch protocol, the XMODEM-1K
block protocol and the YMODEM batch protocol are all supported by 
.IR xmodem .
For details of the protocols,
see the document edited by Chuck Forsberg titled
.I
XMODEM/YMODEM Protocol Reference.
.sp
Option Flags are case insensitive and may be preceded by an optional "-"
character.
.PP
.SH PARAMETERS
Exactly one of the following must be selected:
.TP
.B rb  
Receive Binary - files are placed on the Unix disk without conversion.
.I Xmodem
will silently destroy existing files of the same name.
.TP
.B rt  
Receive Text - files are converted from the CP/M and MS-DOS
format of CR-LF pairs to the Unix convention of newline 
characters only between lines.  
Null bytes are ignored and bit 8 of each character is stripped (which makes 
Wordstar files much more readable).
A CTRL-Z character is deemed to indicate the EOF location in the incoming
file.
The resulting file
is acceptable to the Unix editors and compilers, and is usually slightly
smaller than the original file.
.I Xmodem
will silently destroy existing files of the same name.
.TP
.B sb  
Send Binary - files are sent without conversion as they exist on the Unix disk.
.TP
.B st  
Send Text - newline characters in the file are converted to CR-LF pairs
in accord with the CP/M and MS-DOS conventions for text files.  The file grows
slightly as this occurs so the estimate of file transmission size and
time are always optimistically low.
.PP
.SH OPTIONS
.TP
.B y
Select the YMODEM batch protocol for sending files; a list of files specified
on the command line will be sent in sequence.  The YMODEM batch protocol is 
used automatically for file reception if the sending program requests it.
.TP
.B m
Select the MODEM7 batch protocol for sending files; a list of files specified
on the command line will be sent in sequence.  The MODEM7 batch protocol is 
used automatically for file reception if the sending program requests it.
If this flag is specified for a batch receive, (\fIxmodem rbm\fR, for example),
the transfer starts in the checksum rather mode rather than the CRC mode.
.TP
.B k
Select the XMODEM-1K file transfer mode for sending files. Use of 1K packets on
low-error lines increases throughput.  
However, over direct connections at 9600 bps to a busy host, 1K packets may
cause data overflows generating excessive retries.
1K packets are automatically
used for file reception if the sending program requests it.
If this flag is specified with the YMODEM flag in a batch receive (\fIxmodem
rbYK\fR, for example), the program will attempt to use the "KMD/IMP" convention
to invoke 1K file transfers.
.TP
.B c   
Select the CRC-16 error-checking protocol on receive.  CRC mode is better at catching
transmission errors that occur than the alternative checksum protocol.  
CRC mode is automatically selected for file
transmission if the receiving modem program requests it.
.TP
.B t
Indicates the Unix system is Too Busy and 
.I xmodem
should fall back to a simpler I/O strategy than normal.
.TP
.B d   
Delete the 
.I xmodem.log
file before file transfer is begun.
.TP
.B l   
Do NOT write to the log file.  If logging is selected, a file
.I xmodem.log 
will be created (or appended to), with entries for significant events, errors
and retries.  This can be useful to see why things went wrong
when they do.
.TP
.B x
Toggle on debug mode.  If debug mode is selected, copious and possibly
useful debugging information will be placed in 
.IR xmodem.log .
.SH "FILE NAMES"
Files transmitted using one of the batch modes
will be stored on the remote machine under a CP/M-ified name (limited
to eight characters plus a three character extension; ":" characters will
be turned into "/" characters; all characters will be in monocase).  
Files received using one of the batch modes
will be stored under their transmitted names (except that any "/" characters
in the file name will be converted into ":" characters, all upper-case
characters will be translated into lower case and trailing dots will be
expunged).
.PP
When a batch receive is requested,
.I xmodem
takes a wait and see attitude and can adapt to either batch protocol or even
a classic XMODEM transfer (note that CRC-16 mode is automatically set under
these circumstances unless the b flag is specified).
If a classic, "non-batch" XMODEM file reception takes place, 
the received file is stored as
.IR xmodem.in .
File names present on the command line for a batch receive are ignored.
.SH NOTES
While waiting for the beginning of a file transfer, 
.I xmodem
treats two CAN (CTRL-X) characters that are received within 3 seconds
as a request to abort.  CAN characters will not cause an abort if received
during a file transfer.
.PP
If 10 or more errors are detected during the transmission or reception of any
one packet, the transfer is aborted.
.PP
Squeezed, compressed or ARCed files must be transferred in binary mode, even if they
contain text.
.PP
If you use 
.I xmodem
over a 
.I rlogin
link, you must use the flag
.IR "rlogin machine -8" .
.PP
If an unexpected error occurs before a file is completely received, the
incomplete file is deleted.
.PP
Files received using both binary and text mode in a YMODEM batch transfer 
will be truncated
to the file size specified in the YMODEM header (extra CR characters in the
incoming file are correctly handled).  File sizes are included in
the YMODEM header when sending binary files, 
but are zeroed on transmission of text files (since it would be too slow to
remove CR characters in advance).
File modification times are set for received files if present in the YMODEM
header; they are included in the headers for transmitted files.
.PP
The "KMD/IMP" record count field in the YMODEM header is both set and read.
.SH EXAMPLES
To receive a text file transmitted from a micro (using CRC-16
error-checking) and store it under the
name 
.IR file.name ,
use the command line
.RS
.B "xmodem rtc file.name"
.RE
Note that if the transmitting program on the micro uses the 1K packet
protocol or either batch protocol,
.I xmodem
detects this automatically and takes appropriate action.  Further
note that if one of the batch protocols is used, the received file(s)
will be stored under their own names and the name on the command line
(if any) will be ignored.
.PP
To send a set of text files to a microcomputer using 1K packets and the
YMODEM batch protocol, use the command line
.RS
.B "xmodem styk *.txt"
.RE
.SH FILES
xmodem.log (if logging is enabled)
.SH BUGS
Batch mode could be smarter about bad file-names in the midst of a
batch transmit/receive.
.PP
Batch mode could allow a mixture of binary and text files.
.PP
Bare Carriage Return characters (i.e., those not immediately followed by a
Line Feed character) are mishandled in a received file when using text mode.
A file with ``overstruck'' lines will thus come out looking funny.
.PP
YMODEM header packets are only sent as 128 byte packets.  Thus VERY long file
names will not be handled properly.
.SH SEE ALSO
kermit(1)
.SH AUTHOR
Steve Grandi, National Optical Astronomy Observatories.  Based on
.I xmodem
by Brian Kantor, University of California at San Diego.
This, in turn, was based on
.I umodem
by Lauren Weinstein, Richard Conn and others.


sbbs/news
#! /bin/sh
# news : A bulletin update system

p=$dir/misc/newnews
c=$dir/bulletins
g="`grep \"^$bbs_name:\" $p`"

if [ "$g" != "" ] ; then
  rec="`grep \"^$bbs_name:\" $p `"
  rec="`echo \"$rec\" | awk -F: '{print $2}'`"
 else
  rec=1
  echo "$bbs_name:1" >> $p
fi

b=`ls $c | wc -l`

if [ $rec -gt $b ] ; then
  sleep 5
  exit
fi

echo "===========[System News]==========="

while [ $rec -le $b ] ; do
  cat $c/$rec
  f=/tmp/sbbs.tpb.$$
  $dir/programs/lock /tmp/sbbs.lock
  grep -v "^$bbs_name:" $p > $f
  cat $f > $p
  rm $f
  rc=`expr $rec + 1`
  echo "$bbs_name:$rc" >> $p
  rm /tmp/sbbs.lock
  echo "Continue or Quit) C\b\c"
  read default
  default=`echo $default | cut -c1`
  if [ "$default" = Q -o "$default" = q ] ; then
    exit
  fi
  rec=`expr $rec + 1`
done


xmodem3.4-unix/xmodem.c
/*
 *  XMODEM -- Implements the Christensen XMODEM protocol, 
 *            for packetized file up/downloading.    
 *
 *	I have tried to keep the 4.2isms (select system call, 4.2BSD/v7 tty
 *	structures, gettimeofday system call, etc.) in the source file
 *	getput.c; but I make no guarantees.  Also, I have made no attempt to
 *	keep variable names under 7 characters (although a cursory check
 *	shows that all variables are unique within 7 first characters).
 *	See the README file for some notes on SYS V adaptations.
 *	The program has been successfully run on VAXes (4.3BSD) and SUN-3s
 *	(2.x/3.x) against MEX-PC and ZCOMM/DSZ.
 *
 *   -- Based on UMODEM 3.5 by Lauren Weinstein, Richard Conn, and others.
 *
 *  XMODEM Version 1.0  - by Brian Kantor, UCSD (3/84)
 *
 *  Version 2.0 (CRC-16 and Modem7 batch file transfer) -- Steve Grandi, NOAO (5/85)
 *
 *  Version 2.1 (1K packets) -- Steve Grandi, NOAO (7/85)
 *
 *  Version 2.2 (general clean-ups and multi-character read speed-ups) -- Steve Grandi, NOAO (9/85)
 *
 *  Version 2.3 (napping while reading packet; split into several source files) -- Steve Grandi, NOAO (1/86)
 *
 *  Version 3.0 (Ymodem batch receive; associated changes) -- Steve Grandi, NOAO (2/86)
 *
 *  Version 3.1 (Ymodem batch send; associated changes) -- Steve Grandi, NOAO (8/86)
 *
 *  Version 3.2 (general cleanups) -- Steve Grandi, NOAO (9/86)
 *
 *  Released to the world (1/87)
 *
 *  Version 3.3 (see update.doc) -- Steve Grandi, NOAO (5/87)
 *
 *  Version 3.4 (see update.doc) -- Steve Grandi, NOAO (10/87)
 *
 *  Released to the world (1/88)
 *
 *  Please send bug fixes, additions and comments to:
 *	{ihnp4,hao}!noao!grandi   grandi@noao.arizona.edu
 */

#include "xmodem.h"

main(argc, argv)
int argc;
char **argv;
{
	char *getenv();
	FILE *fopen();
	char *unix_cpm();
	char *strcpy();
	char *strcat();
	
	char *fname = filename;		/* convenient place to stash file names */
	char *logfile = "xmodem.log";	/* Name of LOG File */
	
	char *stamptime();		/* for timestamp */

	char *defname = "xmodem.in";	/* default file name if none given */

	struct stat filestatbuf;	/* file status info */

	int index;
	char flag;
	long expsect;

	/* initialize option flags */

	XMITTYPE = 't';		/* assume text transfer */
	DEBUG = FALSE;		/* keep debugging info in log */
	RECVFLAG = FALSE;	/* not receive */
	SENDFLAG = FALSE;	/* not send either */
	BATCH = FALSE;		/* nor batch */
	CRCMODE = FALSE;	/* use checksums for now */
	DELFLAG = FALSE;	/* don't delete old log file */
	LOGFLAG = TRUE;		/* keep log */
	LONGPACK = FALSE; 	/* do not use long packets on transmit */
	MDM7BAT = FALSE;	/* no MODEM7 batch mode */
	YMDMBAT = FALSE;	/* no YMODEM batch mode */
	TOOBUSY = FALSE;	/* not too busy for sleeping in packet read */

	printf("XMODEM Version %d.%d", VERSION/10, VERSION%10);
	printf(" -- UNIX-Microcomputer Remote File Transfer Facility\n");

	if (argc == 1)
		{
		help();
		exit(-1);
		}

	index = 0;		/* set index for flag loop */

	while ((flag = argv[1][index++]) != '\0')
	    switch (flag) {
		case '-' : break;
		case 'X' :
		case 'x' : DEBUG = TRUE;  /* turn on debugging log */
			   break;
		case 'C' :
		case 'c' : CRCMODE = TRUE; /* enable CRC on receive */
			   break;
		case 'D' :
		case 'd' : DELFLAG = TRUE;  /* delete log file */
			   break;
		case 'L' :
		case 'l' : LOGFLAG = FALSE;  /* turn off log  */
			   break;
		case 'm' :
		case 'M' : MDM7BAT = TRUE;  /* turn on MODEM7 batch protocol */
			   BATCH   = TRUE;
			   break;
		case 'y' :
		case 'Y' : YMDMBAT = TRUE;  /* turn on YMODEM batch protocol */
			   BATCH   = TRUE;
			   break;
		case 'k' :
		case 'K' : LONGPACK = TRUE;  /* use 1K packets on transmit */
			   break;
		case 't' :
		case 'T' : TOOBUSY = TRUE;  /* turn off sleeping */
			   break;
		case 'R' :
		case 'r' : RECVFLAG = TRUE;  /* receive file */
			   XMITTYPE = gettype(argv[1][index++]);  /* get t/b */
			   break;
		case 'S' :
		case 's' : SENDFLAG = TRUE;  /* send file */
			   XMITTYPE = gettype(argv[1][index++]);
			   break;
		default  : printf ("Invalid Flag %c ignored\n", flag);
			   break;
	   }

	if (DEBUG)
		LOGFLAG = TRUE;

	if (LOGFLAG)
	   { 
	     if ((fname = getenv("HOME")) == 0)	/* Get HOME variable */
		error("Fatal - Can't get Environment!", FALSE);
	     fname = strcat(fname, "/");
	     fname = strcat(fname, logfile);
	     if (!DELFLAG)
		LOGFP = fopen(fname, "a");  /* append to LOG file */
	     else
		LOGFP = fopen(fname, "w");  /* new LOG file */
	     if (!LOGFP)
		error("Fatal - Can't Open Log File", FALSE);

	     fprintf(LOGFP,"\n++++++++  %s", stamptime());
	     fprintf(LOGFP,"XMODEM Version %d.%d\n", VERSION/10, VERSION%10);
	     fprintf(LOGFP,"Command line: %s %s", argv[0], argv[1]);
	     for (index=2; index<argc; ++index)
		fprintf(LOGFP, " %s", argv[index]);
	     fprintf(LOGFP, "\n");
	   }

	getspeed();		/* get tty-speed for time estimates */

	if (RECVFLAG && SENDFLAG)
		error("Fatal - Both Send and Receive Functions Specified", FALSE);

	if (MDM7BAT && YMDMBAT)
		error("Fatal - Both YMODEM and MODEM7 Batch Protocols Specified", FALSE);

	if (!RECVFLAG && !SENDFLAG)
		error("Fatal - Either Send or Receive Function must be chosen!",FALSE);
	
	if (SENDFLAG && argc==2)
		error("Fatal - No file specified to send",FALSE);

	if (RECVFLAG && argc==2)
		{
		/* assume we really want CRC-16 in batch, unless we specify MODEM7 mode */ 
		CRCMODE = MDM7BAT ? FALSE : TRUE;
		printf("Ready for BATCH RECEIVE");
		printf(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
		printf("Send several Control-X characters to cancel\n");
		logit("Batch Receive Started");
		logitarg(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
		strcpy(fname, defname);
		}

	if (RECVFLAG && argc>2)
		{
		if(open(argv[2], 0) != -1)  /* check for overwriting */
			{
			logit("Warning -- Target File Exists and is Being Overwritten\n");
			printf("Warning -- Target File Exists and is Being Overwritten\n");
			}
		printf("Ready to RECEIVE File %s", argv[2]);
		printf(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
		printf("Send several Control-X characters to cancel\n");
		logitarg("Receiving in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
		strcpy(fname,argv[2]);
		}

	if (RECVFLAG)
		{  
		setmodes();		/* set tty modes for transfer */

		while(rfile(fname) != FALSE);  /* receive files */

		restoremodes(FALSE);	/* restore normal tty modes */

		sleep(2);		/* give other side time to return to terminal mode */
		exit(0);
		}

	if (SENDFLAG && BATCH) 
		{
		if (YMDMBAT)
			{
			printf("Ready to YMODEM BATCH SEND");
			printf(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
			logit("YMODEM Batch Send Started");
			logitarg(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
			}
		else if (MDM7BAT)
			{
			printf("Ready to MODEM7 BATCH SEND");
			printf(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
			logit("MODEM7 Batch Send Started");
			logitarg(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
			}
		printf("Send several Control-X characters to cancel\n");

		setmodes();
		for (index=2; index<argc; index++) {
			if (stat(argv[index], &filestatbuf) < 0) {
				logitarg("\nFile %s not found\n", argv[index]);
				continue;
			}
			sfile(argv[index]);
		}
		sfile("");
		restoremodes(FALSE);

		logit("Batch Send Complete\n");
		sleep(2);
		exit (0);
		}

	if (SENDFLAG && !BATCH) 
		{
		if (stat(argv[2], &filestatbuf) < 0)
			error("Can't find requested file", FALSE);
		expsect = (filestatbuf.st_size/128)+1;
			
		printf("File %s Ready to SEND", argv[2]);
		printf(" in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");
		printf("Estimated File Size %ldK, %ld Sectors, %ld Bytes\n",
	    	  (filestatbuf.st_size/1024)+1, expsect,
	  	  filestatbuf.st_size);
		projtime(expsect, stdout);
		printf("Send several Control-X characters to cancel\n");
		logitarg("Sending in %s mode\n", (XMITTYPE == 't') ? "text" : "binary");

		setmodes();
		sfile(argv[2]);
		restoremodes(FALSE);

		sleep(2);
		exit(0);
		}
}
